rfcs/0008-storage-ref.md

```diff
@@ -0,0 +1,249 @@
+- Feature Name: Storage References
+- Start Date: 2023-02-10)
+- RFC PR: [FuelLabs/sway-rfcs#0008](https://github.com/FuelLabs/sway-rfcs/pull/023)
+- Sway Issue: [FueLabs/sway#0000](https://github.com/FuelLabs/sway/issues/3043)
+
+# Summary
+
+[summary]: #summary
+
+One paragraph explanation of the feature.
+
+# Motivation
+
+[motivation]: #motivation
+
+Why are we doing this? What use cases does it support? What is the expected outcome?
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+A storage reference stores a `b256` key that describes a storage slot:
+
+```rust
+pub struct StorageRef<T> {
+    key: b256,
+}
+```
+
+We can implement `store` and `get` as follows:
+
+```rust
+impl<T> StorageRef<T> {
+    #[storage(read)]
+    pub fn get(self) -> Option<T> {
+        std::storage::get(self.key)
+    }
+
+    #[storage(write)]
+    pub fn store(self, other: T) {
+        std::storage::store(self.key, other)
+    }
+}
+```
+
+One can create a `StorageRef` manually (it's really just a wrapper around a `b256` key) or via a call to `as_storage_ref` from a new trait called `AsStorageRef`:
+
+```rust
+pub trait AsStorageRef<T> {
+    fn as_storage_ref(self) -> StorageRef<T>;
+}
+```
+
+Calling `as_storage_ref` is only valid for a storage accesses. A storage access looks like `storage.a.b.c...`. For example:
+
+```rust
+struct MyStruct {
+    x: u64,
+    y: b256,
+}
+
+storage {
+    s: MyStruct, 
+}
+
+fn foo() {
+    let x_ref: StorageRef<u64> = storage.s.x.as_storage_ref();
+    let y_ref: StorageRef<b56> = storage.s.y.as_storage_ref();
+    let s_ref: StorageRef<MyStruct> = storage.s.as_storage_ref();
+}
+```
+
+In order to use a storage reference, one can call `store` and `get` as follows:
+```rust
+let x_ref: StorageRef<u64> = storage.s.x.as_storage_ref();
+x_ref.store(42); // equivalent to `storage.x = 42`
+let x = x_ref.get(); // equivalent to `let x = storage.s.x`
```

Comment:
This `get()` isn't really equivalent as the comments suggest, as `get()` in the proposed API returns `Option<T>`.

Reply:
That's true.. this is an oversight as `get` was changed recently to return an `Option`.